### ISSUE TITLE
buildGoModule doesn't correctly parse CFLAGS from dependencies

### ISSUE THREAD

**Issue by bergkvist** [NixOS/nixpkgs#221375]
### Describe the bug
The following builds fine with `nix-shell -p go --run "go build"` after generating `go.mod` and `go.sum`.
```go
package main

import (
	"fmt"

	blst "github.com/supranational/blst/bindings/go"
)

func main() {
	agg := new(blst.P1Aggregate)
	fmt.Println(agg)
}
```

However, using [`buildGoModule`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/go/module.nix), we get an include error:
```
vendor/github.com/supranational/blst/bindings/go/blst.go:16:11: fatal error: 'blst.h' file not found
 #include "blst.h"
          ^~~~~~~~
1 error generated.
```

To see where this is coming from, see:
https://github.com/supranational/blst/blob/a7fd1f584d26b0ae6cdc427976ea1d8980f7e15d/bindings/go/blst.go#L13-L16
```
// #cgo CFLAGS: -I${SRCDIR}/.. -I${SRCDIR}/../../build -I${SRCDIR}/../../src -D__BLST_CGO__ -fno-builtin-memcpy -fno-builtin-memset
// #cgo amd64 CFLAGS: -D__ADX__ -mno-avx
// #cgo mips64 mips64le ppc64 ppc64le riscv64 s390x CFLAGS: -D__BLST_NO_ASM__
// #include "blst.h"
```

Here, `blst.h` is found in `${SRCDIR}/..` when doing a normal `go build` from the terminal. For some reason this doesn't work when using `buildGoModule`.

### Steps To Reproduce
I made an example repo where I pinned nixpkgs here: https://github.com/bergkvist/cflags-buildgomodule

The example is reproducible on both macOS (aarch64) and Linux (x86_64)

```nix
{ pkgs ? import (builtins.fetchTarball {
    url = "https://github.com/NixOS/nixpkgs/archive/a7cc81913bb3cd1ef05ed0ece048b773e1839e51.tar.gz";
    sha256 = "0mzqw20rj0i0wi8zbdszcm81zadmyk1mspi70bf7qkfdv201fl28";
  }) {}
}:
pkgs.buildGoModule {
  name = "cflags-buildgomodule";
  src = builtins.path {
    name = "cflags-buildgomodule";
    path = ./.;
    filter = path: type: builtins.elem (baseNameOf path) [
      "main.go"
      "go.mod"
      "go.sum"
    ];
  };
  vendorSha256 = "sha256-VJj1itAxFDBzGBWVW69FKWireAiR2G1kvdEpfHqBRiQ=";
}
```

**Comment by qbit** [comment]
It should work if you add `proxyVendor = true;` and re-create the `vendorHash`.

Bulrb from the module.. module ( :D ): 
> Whether to fetch (go mod download) and proxy the vendor directory.
> This is useful if your code depends on c code and go mod tidy does not
> include the needed sources to build or if any dependency has case-insensitive
> conflicts which will produce platform dependant `vendorHash` checksums.

**Comment by bergkvist** [comment]
Thanks @qbit! This did indeed work!

In case someone else stumbles upon this: Last time I tested out `proxyVendor = true;`, I had also set the `GOPRIVATE` environment variable, which caused `GONOPROXY` to be set to the same value - which again caused the private/cached repos to be ignored by the `file://`-proxy and not be found. Removing the `GOPRIVATE` environment variable allowed me to successfully build.
